### ISSUE TITLE
FASTBuild 1.13 gives false "Missing precompiled header option Yc in PCHOptions" error when Yc is present

### ISSUE THREAD

**Issue by john-blackburn** [fastbuild/fastbuild#1039]
FASTBuild v 1.13 is failing to parse my `fbuild.bff` file, claiming that my `PCHOptions` string does not contain `Yc` even when `Yc` is present. Previous versions of FASTBuild worked just fine. The error occurs when the `PCHOptions` string contains `\"` in which the backslash seems to be getting treated as an escape character when it searches for the `Yc` (so it doesn't find it). This is not correct : caret is the correct escape character. Here is an example BFF that shows the bug:

```
.msvcPath='C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.35.32215\bin\Hostx64\x64'

.Compiler ='$msvcPath$\cl.exe'

.CompilerOptions='/nologo /c /Fo"%2" "%1" /Fpobj\precompiled.pch \I"my\include\path\"'
.PCHInputFile = 'precompiled.cpp'
.PCHOutputFile = 'obj\precompiled.pch'

.PCHOptions = .CompilerOptions - '/Fo"%2"' + ' /Fo"%3"' + ' /Yc"precompiled.hpp"' 

Print(.PCHOptions)

.CompilerOptions + ' /Yu"precompiled.hpp"'

ObjectList('hello_obj')
{
     .CompilerInputPath = '.' 
     .CompilerOutputPath = 'obj'
}
```
If you run this with fbuild 1.13 it fails to parse with 
```
C:\Users\john.blackburn\Documents\fbuild_test\pch_test\fbuild.bff(16,1): FASTBuild Error #1302 - ObjectList() - Missing Precompiled Header option 'Yc' in 'PCHOptions'.
ObjectList('hello_obj')
^
\--here
```

But `Yc` is present in `PCHOptions` as can be seen from the above `Print` statement which outputs:
```
.PCHOptions = '/nologo /c  "%1" /Fpobj\precompiled.pch \I"my\include\path\" /Fo"%3" /Yc"precompiled.hpp"'
```

The problem seems to be the `\"` at the end of `\I"my\include\path\"` (it succeeds if that backslash is removed). The parser seems to give up there and so doesn't notice the `Yc` at the end of the string. This script works fine in all versions of FASTBuild previous to 1.13.

**Comment by ffulin** [comment]
The TL;DR: here is that your args are (and have always been) incorrect due to the backslash escaping the closing quote and this was covered up by incorrect behavior in the previous version of FASTBuild that was fixed in v1.13.

**Details:**
FASTBuild did not correctly manage param quoting and escape backslashes, and was entirely inconsistent in various ways between Windows and OSX/Linux. To fix some various long standing bugs caused by this, it was necessary to make this behavior consistent and this unfortunately means that some existing "bad" bff files will now surface the errors they ideally should have all along.

At play in this case is specifically the backslashes acting as an escape character that you noticed. As of v1.13, this behaves the same way as the shell/CreateProcess etc.

So, specifically, in your example, the args:
```
/nologo /c  "%1" /Fpobj\precompiled.pch \I"my\include\path\" /Fo"%3" /Yc"precompiled.hpp"
```
Would actually be seen by the cl.exe process as from cmd.exe as:
```
/nologo
/c 
%1
/Fpobj\precompiled.pch
\Imy\include\path" /Fo%3 /Ycprecompiled.hpp
```
and now FASTBuild sees them the same way, which in this instance also means it detects that /Yc is not correctly set (though that's only one of several problems those args if there are other args that follow)

To fix your bff, it is necessary to either remove the backslash before the quote:
```
/Fpobj\precompiled.pch \I"my\include\path"
```
or escape it with another slash:
```
/Fpobj\precompiled.pch \I"my\include\path\\"
```

To confirm this is the correct behavior, you can perform some tests entirely outside of FASTBuild too with the following example.
Given a file test.cpp:
```
#if !defined(TEST)
    #error Missing TEST define
#endif
```
compiling from cmd.exe with:
```
cl.exe -c -nologo test.cpp /I"Folder\" -DTEST
```
results in failure as:
```
test.cpp
test.cpp(3): fatal error C1189: #error:  Missing TEST define
```
while this:
```
cl.exe -c -nologo test.cpp /I"Folder" -DTEST
```
or this:
```
cl.exe -c -nologo test.cpp /I"Folder\\" -DTEST
```
results in success.

Breaking changes to FASTBuild are taken very seriously and avoided unless absolutely necessary. In this case, the only sane way to fix the bugs was to correct the long-standing incorrect behavior and to try and call this change out in the release notes.

Please let me know if I can further clarify or check any of the behavior.

**Comment by ffulin** [comment]
I'm going to close this issue as I've answered as best I can. Please let me know if you have any followup questions or if you think I've overlooked something.
